runtime,agent: Add SELinux support for containers inside the guest

ManaSugi opened the pull request:
Kata containers enable SELinux for containers running inside the guest
to improve security. With this feature enabled, `kata-agent` is launched
with the `container_runtime_t` type and container processes are launched
with `container_t` type. To enable SELinux in the guest, SELinux must also
be enabled on the host side. This feature is disabled by default.

Fixes: #4812

Signed-off-by: Manabu Sugimoto <Manabu.Sugimoto@sony.com>

ManaSugi commented:
# Experiments with containerd

## Requirements

* Enable SELinux support on containerd

`/etc/containerd/config.toml`
```toml
[plugins]
  [plugins.cri]
    enable_selinux = true
```

* Make sure that SELinux is enabled on the host

```sh
$ getenforce
Permissive
```

* Enable SELinux in the runtime configuration

```sh
$ sudo sed -i '/^disable_guest_selinux/ s/true/false/' /etc/kata-containers/configuration.toml
$ sudo sed -i -e 's/^kernel_params = "\(.*\)"/kernel_params = "\1 security=selinux"/g' /etc/kata-containers/configuration.toml
$ sudo sed -i -e 's/^virtio_fs_extra_args = \[\(.*\)\]/virtio_fs_extra_args = \[\1, "-o", "xattr"\]/g' /etc/kata-containers/configuration.toml
```

* Create rootfs for SELinux

```sh
$ script -fec 'sudo -E GOPATH=$GOPATH USE_DOCKER=true SELINUX=yes ./rootfs.sh centos'
```

* Build rootfs for SELinux

To confirm that the processes inside the VM are labeld properly, `ps` command is required, so add the `procps` package for debugging as follows.

`kata-containers/tools/osbuilder/rootfs-builder/centos/config.sh`
```sh
[ "$SELINUX" = yes ] && PACKAGES+=" selinux-policy-targeted container-selinux policycoreutils procps"
```

```sh
$ script -fec 'sudo -E USE_DOCKER=true SELINUX=yes ./image_builder.sh ${ROOTFS_DIR}'
```

* Compile the guest kernel with SELinux enabled

```
CONFIG_SECURITY_SELINUX=y
CONFIG_SECURITY_SELINUX_DEVELOP=y (optional)
```

## Launch Kata container with SELinux

### 1. Launch a container using `crictl`

```sh
$ uname -a
Linux xxx 5.10.0-1057 #61-Ubuntu SMP Thu Jan 13 15:06:11 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

$ sudo crictl run -r kata container.json sandbox.json
...
DEBU[0000] RunPodSandboxResponse: 44dda92f5b49e8e95c79ce651490619b3abcf7acde0f9c72cae370e20e7c24ff
...
DEBU[0002] CreateContainerResponse: 1228bdb66afebcde783a7b440956de99bca1601753e1d28136722d89e6a44877
```

<details><summary>container.json</summary><div>

```json
{
    "metadata": {
        "name": "kata-ubuntu"
    },
    "image": {
        "image": "docker.io/ubuntu"
    },
    "command": [
        "sh"
    ],
    "envs": [
        {
            "key": "PATH",
            "value": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        },
        {
            "key": "TERM",
            "value": "xterm"
        }
    ],
    "log_path": "busybox.0.log",
    "stdin": true,
    "stdin_once": true,
    "tty": true,
    "linux": {
    }
}
```
</div></details>
<details><summary>sandbox.json</summary><div>

```json
{
    "metadata": {
        "name": "kata-sandbox",
        "namespace": "default",
        "uid": "kata-sandbox"
    },
    "linux": {
    }
}
```
</div></details>

### 2. Attach the container

We can confirm that all the files and processes are labled for the containre security policy which is provided by [container-selinux](https://github.com/containers/container-selinux).

```
$ sudo crictl attach --stdin --tty 1228bdb66afebcde783a7b440956de99bca1601753e1d28136722d89e6a44877

# ls -lZ
total 52
lrwxrwxrwx.  1 root root system_u:object_r:container_file_t:s0:c635,c637    7 Aug  1 13:22 bin -> usr/bin
drwxr-xr-x.  2 root root system_u:object_r:container_file_t:s0:c635,c637 4096 Apr 18 10:28 boot
drwxr-xr-x.  5 root root system_u:object_r:container_file_t:s0:c635,c637  320 Aug  3 11:08 dev
drwxr-xr-x.  1 root root system_u:object_r:container_file_t:s0:c635,c637 4096 Aug  1 13:25 etc
drwxr-xr-x.  2 root root system_u:object_r:container_file_t:s0:c635,c637 4096 Apr 18 10:28 home
lrwxrwxrwx.  1 root root system_u:object_r:container_file_t:s0:c635,c637    7 Aug  1 13:22 lib -> usr/lib
lrwxrwxrwx.  1 root root system_u:object_r:container_file_t:s0:c635,c637    9 Aug  1 13:22 lib32 -> usr/lib32
lrwxrwxrwx.  1 root root system_u:object_r:container_file_t:s0:c635,c637    9 Aug  1 13:22 lib64 -> usr/lib64
lrwxrwxrwx.  1 root root system_u:object_r:container_file_t:s0:c635,c637   10 Aug  1 13:22 libx32 -> usr/libx32
drwxr-xr-x.  2 root root system_u:object_r:container_file_t:s0:c635,c637 4096 Aug  1 13:22 media
drwxr-xr-x.  2 root root system_u:object_r:container_file_t:s0:c635,c637 4096 Aug  1 13:22 mnt
drwxr-xr-x.  2 root root system_u:object_r:container_file_t:s0:c635,c637 4096 Aug  1 13:22 opt
dr-xr-xr-x. 96 root root system_u:object_r:proc_t:s0                        0 Aug  3 11:08 proc
drwx------.  2 root root system_u:object_r:container_file_t:s0:c635,c637 4096 Aug  1 13:25 root
drwxr-xr-x.  5 root root system_u:object_r:container_file_t:s0:c635,c637 4096 Aug  1 13:25 run
lrwxrwxrwx.  1 root root system_u:object_r:container_file_t:s0:c635,c637    8 Aug  1 13:22 sbin -> usr/sbin
drwxr-xr-x.  2 root root system_u:object_r:container_file_t:s0:c635,c637 4096 Aug  1 13:22 srv
dr-xr-xr-x. 11 root root system_u:object_r:sysfs_t:s0                       0 Aug  3 11:07 sys
drwxrwxrwt.  2 root root system_u:object_r:container_file_t:s0:c635,c637 4096 Aug  1 13:25 tmp
drwxr-xr-x. 14 root root system_u:object_r:container_file_t:s0:c635,c637 4096 Aug  1 13:22 usr
drwxr-xr-x. 11 root root system_u:object_r:container_file_t:s0:c635,c637 4096 Aug  1 13:25 var

# ps auxZ
LABEL                           USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
system_u:system_r:container_t:s0:c635,c637 65535 1 0.0  0.0   992     4 ?        S    11:07   0:00 /pause
system_u:system_r:container_t:s0:c635,c637 root 2 0.0  0.0   2884    92 ?        Ss   11:08   0:00 sh
system_u:system_r:container_t:s0:c635,c637 root 4 0.0  0.1   7836  2952 ?        R+   11:10   0:00 ps auxZ
```

### 3. Attach the VM

We can confirm that the `kata-agent` is labeld as `container_runtime_t` and the container processes such as `/pause` and `sh` are labeled as `container_t`. There are no error messages about `audit` in `dmesg`.

```
$ sudo kata-runtime exec 44dda92f5b49e8e95c79ce651490619b3abcf7acde0f9c72cae370e20e7c24ff
bash-5.1# Linux localhost 5.15.26 #8 SMP Sun Jul 10 21:59:13 JST 2022 x86_64 x86_64 x86_64 GNU/Linux

bash-5.1# getenforce
Enforcing

bash-5.1# ps auxZ
...
system_u:system_r:kernel_t:s0   root          40  0.0  0.0      0     0 ?        I<   11:07   0:00 [ext4-rsv-conver]
system_u:system_r:container_runtime_t:s0 root 63  0.0  0.0  18560  1072 ?        Ssl  11:07   0:00 /usr/bin/kata-agent
system_u:system_r:kernel_t:s0   root          64  0.0  0.0      0     0 ?        I    11:07   0:00 [kworker/u48:3-events_unbound]
system_u:system_r:chronyd_t:s0  chrony        71  0.0  0.0  10280   536 ?        S    11:07   0:00 /usr/sbin/chronyd -F 2
system_u:system_r:container_t:s0:c635,c637 65535 73 0.0  0.0  992     4 ?        S    11:07   0:00 /pause
system_u:system_r:container_t:s0:c635,c637 root 75 0.0  0.0  2884    92 pts/0    Ss+  11:08   0:00 sh
system_u:system_r:container_runtime_t:s0 root 78  0.0  0.0   4724   488 pts/1    Ss   11:15   0:00 [bash]
system_u:system_r:container_runtime_t:s0 root 80  0.0  0.0  15476   864 pts/1    R+   11:15   0:00 ps auxZ

bash-5.1# ls -laZ
total 140
drwxr-xr-x. 19 root root system_u:object_r:root_t:s0        4096 Aug  3 03:58 .
drwxr-xr-x. 19 root root system_u:object_r:root_t:s0        4096 Aug  3 03:58 ..
dr-xr-xr-x.  2 root root system_u:object_r:mnt_t:s0         4096 Aug  9  2021 afs
lrwxrwxrwx.  1 root root system_u:object_r:bin_t:s0            7 Aug  9  2021 bin -> usr/bin
dr-xr-xr-x.  2 root root system_u:object_r:boot_t:s0        4096 Aug  9  2021 boot
drwxr-xr-x.  8 root root system_u:object_r:device_t:s0      2820 Aug  3 11:07 dev
drwxr-xr-x. 45 root root system_u:object_r:etc_t:s0         4096 Aug  3 03:58 etc
drwxr-xr-x.  2 root root system_u:object_r:home_root_t:s0   4096 Aug  9  2021 home
lrwxrwxrwx.  1 root root system_u:object_r:lib_t:s0            7 Aug  9  2021 lib -> usr/lib
lrwxrwxrwx.  1 root root system_u:object_r:lib_t:s0            9 Aug  9  2021 lib64 -> usr/lib64
drwx------.  2 root root system_u:object_r:lost_found_t:s0 16384 Aug  3 03:58 lost+found
drwxr-xr-x.  2 root root system_u:object_r:mnt_t:s0         4096 Aug  9  2021 media
drwxr-xr-x.  2 root root system_u:object_r:mnt_t:s0         4096 Aug  9  2021 mnt
drwxr-xr-x.  2 root root system_u:object_r:usr_t:s0         4096 Aug  9  2021 opt
dr-xr-xr-x. 94 root root system_u:object_r:proc_t:s0           0 Aug  3 11:07 proc
dr-xr-x---.  2 root root system_u:object_r:admin_home_t:s0  4096 Aug  9  2021 root
drwxr-xr-x.  9 root root system_u:object_r:var_run_t:s0      200 Aug  3 11:07 run
lrwxrwxrwx.  1 root root system_u:object_r:bin_t:s0            8 Aug  9  2021 sbin -> usr/sbin
drwxr-xr-x.  2 root root system_u:object_r:var_t:s0         4096 Aug  9  2021 srv
dr-xr-xr-x. 11 root root system_u:object_r:sysfs_t:s0          0 Aug  3 11:07 sys
drwxrwxrwt.  4 root root system_u:object_r:tmpfs_t:s0         80 Aug  3 11:07 tmp
drwxr-xr-x. 12 root root system_u:object_r:usr_t:s0         4096 Aug  3 03:55 usr
drwxr-xr-x. 15 root root system_u:object_r:var_t:s0         4096 Aug  3 03:58 var

bash-5.1# dmesg | grep audit
[    0.042669] calling  audit_init+0x0/0x16a @ 1
[    0.042672] audit: initializing netlink subsys (disabled)
[    0.042703] initcall audit_init+0x0/0x16a returned 0 after 0 usecs
[    0.044007] audit: type=2000 audit(1659524879.155:1): state=initialized audit_enabled=0 res=1
[    0.093047] calling  audit_classes_init+0x0/0x55 @ 1
[    0.093049] initcall audit_classes_init+0x0/0x55 returned 0 after 1 usecs
[    0.093183] calling  audit_watch_init+0x0/0x35 @ 1
[    0.093185] initcall audit_watch_init+0x0/0x35 returned 0 after 0 usecs
[    0.093186] calling  audit_fsnotify_init+0x0/0x35 @ 1
[    0.093188] initcall audit_fsnotify_init+0x0/0x35 returned 0 after 0 usecs
[    0.093189] calling  audit_tree_init+0x0/0x69 @ 1
[    0.093192] initcall audit_tree_init+0x0/0x69 returned 0 after 1 usecs
[    0.164831] audit: type=1404 audit(1659524879.279:2): enforcing=1 old_enforcing=0 auid=4294967295 ses=4294967295 enabled=1 old-enabled=1 lsm=selinux res=1
[    0.194800] audit: type=1403 audit(1659524879.307:3): auid=4294967295 ses=4294967295 lsm=selinux res=1
[    0.260369] systemd[1]: unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/systemd-journald-audit.socket
```

fidencio commented:
@ManaSugi, there are some conflicts that need to be solved, but in general I must say that I **really like** this PR!

```
CONFIG_SECURITY_SELINUX=y
CONFIG_SECURITY_SELINUX_DEVELOP=y (optional)
```
About the configuration, should we add this by default to our Kernel? (I think so, but I'd like to hear your opinion here).

I'll take some time next week to play with this PR.

ManaSugi commented:
@fidencio Thank you for the comment.

> About the configuration, should we add this by default to our Kernel? (I think so, but I'd like to hear your opinion here).

I was thinking that we could add SELinux kernel config by default, but it dpends on many other configurations, so I'm concerned about the kernel size. For example, `CONFIG_AUDIT` is required for SELinux. If there are no problems about adding the following configurations, I'll add them by default. 

Minimal configurations for SELinux which must be turned on from the default config:
```
# Linux/x86 5.15.26 Kernel Configuration
CONFIG_AUDIT=y
CONFIG_AUDITSYSCALL=y
CONFIG_NETWORK_SECMARK=y
CONFIG_SECURITY_NETWORK=y
CONFIG_LSM_MMAP_MIN_ADDR=65536
CONFIG_SECURITY_SELINUX=y
CONFIG_SECURITY_SELINUX_DEVELOP=y (optional but this would be helpful)
CONFIG_SECURITY_SELINUX_CHECKREQPROT_VALUE=0
CONFIG_SECURITY_SELINUX_SIDTAB_HASH_BITS=9
CONFIG_SECURITY_SELINUX_SID2STR_CACHE_SIZE=256
```

> Also, I'm not exactly sure if we should tie the host and guest SELinux enablement, but I need to carefully think about this.

Yes, I'd like to discuss it. Actually, I have an idea of way for running the guest with enabled SELinux without the host SELinux enablement, so I'll describe it with a figure later. But still, regarding labeling a guest rootfs in advance, I think we need to enable SELinux on the host.

liubin commented on tools/osbuilder/image-builder/image_builder.sh:
```diff
@@ -402,6 +419,27 @@ create_rootfs_image() {
 	cp -a "${rootfs}"/* "${mount_dir}"
+
+	if [ "${SELINUX}" == "yes" ]; then
+		if [ "${AGENT_INIT}" == "yes" ]; then
```
I'm not sure if `AGENT_INIT = yes ` equals using an initrd image.

ManaSugi replied:
Sorry, I made a mistake. It's actually is that guest SELinux does not work with the agent init (`AGENT_INIT=yes`), so I fixed the comment and documentation.

liubin commented on src/agent/rustjail/src/selinux.rs:
```diff
@@ -0,0 +1,74 @@
+    let mut attr_path = Path::new("/proc/thread-self/attr/exec");
+    if !attr_path.exists() {
+        // Fall back to the old convention
+        attr_path = Path::new("/proc/self/task");
```
Is `/proc/self/task` only the prefix of the path?

ManaSugi replied:
Thank you. That was a mistake. I fixed it

liubin commented on src/agent/rustjail/src/validator.rs:
```diff
@@ -82,6 +83,17 @@ fn hostname(oci: &Spec) -> Result<()> {
+
+    if let Some(ref process) = oci.process {
+        if !process.selinux_label.is_empty() && !label_regex.is_match(&process.selinux_label) {
+            return Err(einval());
```
Suggestion adds more detailed message in error message than only one error code.

amshinde replied:
Would suggest the same for the remaining error messages as well.

ManaSugi replied:
Thank you.  I fixed them with detailed error messages.

amshinde commented on src/runtime/config/configuration-acrn.toml.in:
```diff
@@ -189,6 +189,21 @@ internetworking_model="@DEFNETWORKMODEL_ACRN@"
+# (format: "user:role:type")
+# Note: You cannot specify MLS policy with the label because the sensitivity levels and
+# categories are determined automatically by high-level container runtimes such as containerd.
+#guest_selinux_label="@DEFGUESTSELINUXLABEL@"
```
I am not sure if we want to provide this as a config option at all, I dont see the need for the end-user changing this. We can add this at a later point if a specific need arises.

ManaSugi replied:
Yes, as I mentioned in #4812, I'm also not sure if the config is helpful for Kata users. 
The users cannot modify SELinux security policy for containers through the [Kubernetes' security context](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/#selinuxoptions-v1-core), so this config will be an alternative way for modifying the policy. But I agree with your idea that adds this at alter point when a specific need arises.

> 3. Allow users to apply a custom SELinux security policy to a container
As explained above, the security policy is overridden for VMM processes. Hence, users cannot set a custom policy other than `container_t` through Kubernetes's configuration. To solve the issue, we add `guest_selinux_label` option in the runtime configuration. Users can apply a custom policy using the configuration except for MCS. This feature is helpful for users who want to loosen restrictions, but I don't know actually if any users want to apply custom policies to a container other than `container_t`.

amshinde commented on src/runtime/virtcontainers/kata_agent.go:
```diff
@@ -911,10 +915,41 @@ func (k *kataAgent) constrainGRPCSpec(grpcSpec *grpc.Spec, passSeccomp bool, str
+			label = defaultSeLinuxContainerLabel
+		}
+
+		mountContext, err := selinux.NewContext(grpcSpec.Linux.MountLabel)
```
Are the grpcSpec.Linux.MountLabel and grpcSpec.Process.SelinuxLabel ordinarily the same?
Feels like the need to have an additional field in the OCI spec to accomodate both the Process labels for the hypervisor and container process, if we were to do this cleanly in the long run.

ManaSugi replied:
Sorry I misunderstood. We just need to change the type from `container_kvm_t` to `container_t`, so the mount label is no longer used. I updated the description about this in #4812.

> 2. Apply a security policy to a container
In the case of general containers, not VM-based containers, high-level container runtimes such as containerd label the security policy for containers (`container_file_t`, etc) with Multi-Category Security (MCS) on the container image and low-level container runtimes such as runc apply the policy (`container_t`, etc) with the same MCS to the container process.
> If we can pass the security policy provided the high-level runtime to the guest, the kata-agent can apply the policy to the containers inside the guest easily. However, this needs an additional work because the SELinux type is overwritten as the new type (`container_kvm_t`) for VMM processes by the high-level runtime and the runtime passes the type to the kata-runtime. This behavior was added by [this commit](https://github.com/containerd/containerd/commit/d715d009061edf5ed0da5aa81fe7b6d2a6b3c10c) to add SELinux support for VMM processes by the https://github.com/kata-containers/runtime/pull/2503. Hence, we need to go back from the `container_kvm_t` to the `container_t` on the kata-runtime side.

amshinde commented on tools/osbuilder/rootfs-builder/rootfs.sh:
```diff
@@ -391,17 +397,8 @@ build_rootfs_distro()
 		engine_run_args+=" $(docker_extra_args $distro)"
 
-		# Relabel volumes so SELinux allows access (see docker-run(1))
-		if command -v selinuxenabled > /dev/null && selinuxenabled ; then
```
Why is this removed?

ManaSugi replied:
Sorry, I made a mistakes. This is still needed when we use `USE_DOCKER`, so I restored this.

amshinde commented:
@ManaSugi Nice work! Would suggest to split the PR into separate commits to make it easier to digest and review.

> About the configuration, should we add this by default to our Kernel? (I think so, but I'd like to hear your opinion here).]

@ManaSugi Can you measure by how much the kernel size increases in memory when you add those configs?

ManaSugi commented on tools/osbuilder/image-builder/image_builder.sh:
```diff
@@ -170,12 +185,14 @@ build_with_container() {
 		   -v "${script_dir}/../scripts":"/scripts" \
 		   -v "${rootfs}":"/rootfs" \
 		   -v "${image_dir}":"/image" \
+		   ${engine_run_args} \
```
Thank you. I use `selinuxfs` instead of `engine_run_args`.

ManaSugi commented:
@amshinde @liubin Thank you for the review.

> Would suggest to split the PR into separate commits to make it easier to digest and review.

I split the commit by subsystem.

> Can you measure by how much the kernel size increases in memory when you add those configs?

I measured with the guest kernel `5.15.48` (default kata config version: `94`).
Since the increased kernel size and memory usage were trivial, I added the config by default.

|    |  vmlinux [MB]  | vmlinuz [MB] | memory usage of kernel by dmesg [MB] |
| ---- | ---- | ---- | ---- |
|  Default config [1] | 38.8 | 5.82 | 2034132K/2096628K available (10243K kernel code, 1018K rwdata, 1548K rodata, 972K init, 1288K bss, 62240K reserved) |
|  SELinux config (disabled by kernel cmdline) [2] |  39.4  | 5.96 | 2034132K/2096628K available (10243K kernel code, 1046K rwdata, 1596K rodata, 976K init, 1184K bss, 62240K reserved) |

The configuration used in the evaluation is below.

<details><summary>configuration.toml</summary>

```
# Copyright (c) 2017-2019 Intel Corporation
# Copyright (c) 2021 Adobe Inc.
#
# SPDX-License-Identifier: Apache-2.0
#

# XXX: WARNING: this file is auto-generated.
# XXX:
# XXX: Source file: "config/configuration-qemu.toml.in"
# XXX: Project:
# XXX:   Name: Kata Containers
# XXX:   Type: kata

[hypervisor.qemu]
path = "/usr/bin/qemu-system-x86_64"
kernel = "/usr/share/kata-containers/vmlinux.container"
image = "/usr/share/kata-containers/kata-containers.img"
machine_type = "q35"

# Enable confidential guest support.
# Toggling that setting may trigger different hardware features, ranging
# from memory encryption to both memory and CPU-state encryption and integrity.
# The Kata Containers runtime dynamically detects the available feature set and
# aims at enabling the largest possible one, returning an error if none is
# available, or none is supported by the hypervisor.
#
# Known limitations:
# * Does not work by design:
#   - CPU Hotplug 
#   - Memory Hotplug
#   - NVDIMM devices
#
# Default false
# confidential_guest = true

# Enable running QEMU VMM as a non-root user.
# By default QEMU VMM run as root. When this is set to true, QEMU VMM process runs as
# a non-root random user. See documentation for the limitations of this mode.
# rootless = true

# List of valid annotation names for the hypervisor
# Each member of the list is a regular expression, which is the base name
# of the annotation, e.g. "path" for io.katacontainers.config.hypervisor.path"
enable_annotations = ["enable_iommu"]

# List of valid annotations values for the hypervisor
# Each member of the list is a path pattern as described by glob(3).
# The default if not set is empty (all annotations rejected.)
# Your distribution recommends: ["/usr/bin/qemu-system-x86_64"]
valid_hypervisor_paths = ["/usr/bin/qemu-system-x86_64"]

# Optional space-separated list of options to pass to the guest kernel.
# For example, use `kernel_params = "vsyscall=emulate"` if you are having
# trouble running pre-2.15 glibc.
#
# WARNING: - any parameter specified here will take priority over the default
# parameter value of the same name used to start the virtual machine.
# Do not set values here unless you understand the impact of doing so as you
# may stop the virtual machine from booting.
# To see the list of default parameters, enable hypervisor debug, create a
# container and look for 'default-kernel-parameters' log entries.
kernel_params = " selinux=0 agent.log=debug initcall_debug"

# Path to the firmware.
# If you want that qemu uses the default firmware leave this option empty
firmware = ""

# Path to the firmware volume.
# firmware TDVF or OVMF can be split into FIRMWARE_VARS.fd (UEFI variables
# as configuration) and FIRMWARE_CODE.fd (UEFI program image). UEFI variables
# can be customized per each user while UEFI code is kept same.
firmware_volume = ""

# Machine accelerators
# comma-separated list of machine accelerators to pass to the hypervisor.
# For example, `machine_accelerators = "nosmm,nosmbus,nosata,nopit,static-prt,nofw"`
machine_accelerators=""

# Qemu seccomp sandbox feature
# comma-separated list of seccomp sandbox features to control the syscall access.
# For example, `seccompsandbox= "on,obsolete=deny,spawn=deny,resourcecontrol=deny"`
# Note: "elevateprivileges=deny" doesn't work with daemonize option, so it's removed from the seccomp sandbox
# Another note: enabling this feature may reduce performance, you may enable
# /proc/sys/net/core/bpf_jit_enable to reduce the impact. see https://man7.org/linux/man-pages/man8/bpfc.8.html
#seccompsandbox="on,obsolete=deny,spawn=deny,resourcecontrol=deny"

# CPU features
# comma-separated list of cpu features to pass to the cpu
# For example, `cpu_features = "pmu=off,vmx=off"
cpu_features="pmu=off"

# Default number of vCPUs per SB/VM:
# unspecified or 0                --> will be set to 1
# < 0                             --> will be set to the actual number of physical cores
# > 0 <= number of physical cores --> will be set to the specified number
# > number of physical cores      --> will be set to the actual number of physical cores
default_vcpus = 1

# Default maximum number of vCPUs per SB/VM:
# unspecified or == 0             --> will be set to the actual number of physical cores or to the maximum number
#                                     of vCPUs supported by KVM if that number is exceeded
# > 0 <= number of physical cores --> will be set to the specified number
# > number of physical cores      --> will be set to the actual number of physical cores or to the maximum number
#                                     of vCPUs supported by KVM if that number is exceeded
# WARNING: Depending of the architecture, the maximum number of vCPUs supported by KVM is used when
# the actual number of physical cores is greater than it.
# WARNING: Be aware that this value impacts the virtual machine's memory footprint and CPU
# the hotplug functionality. For example, `default_maxvcpus = 240` specifies that until 240 vCPUs
# can be added to a SB/VM, but the memory footprint will be big. Another example, with
# `default_maxvcpus = 8` the memory footprint will be small, but 8 will be the maximum number of
# vCPUs supported by the SB/VM. In general, we recommend that you do not edit this variable,
# unless you know what are you doing.
# NOTICE: on arm platform with gicv2 interrupt controller, set it to 8.
default_maxvcpus = 0

# Bridges can be used to hot plug devices.
# Limitations:
# * Currently only pci bridges are supported
# * Until 30 devices per bridge can be hot plugged.
# * Until 5 PCI bridges can be cold plugged per VM.
#   This limitation could be a bug in qemu or in the kernel
# Default number of bridges per SB/VM:
# unspecified or 0   --> will be set to 1
# > 1 <= 5           --> will be set to the specified number
# > 5                --> will be set to 5
default_bridges = 1

# Default memory size in MiB for SB/VM.
# If unspecified then it will be set 2048 MiB.
default_memory = 2048
#
# Default memory slots per SB/VM.
# If unspecified then it will be set 10.
# This is will determine the times that memory will be hotadded to sandbox/VM.
#memory_slots = 10

# Default maximum memory in MiB per SB / VM
# unspecified or == 0           --> will be set to the actual amount of physical RAM
# > 0 <= amount of physical RAM --> will be set to the specified number
# > amount of physical RAM      --> will be set to the actual amount of physical RAM
default_maxmemory = 0

# The size in MiB will be plused to max memory of hypervisor.
# It is the memory address space for the NVDIMM devie.
# If set block storage driver (block_device_driver) to "nvdimm",
# should set memory_offset to the size of block device.
# Default 0
#memory_offset = 0

# Specifies virtio-mem will be enabled or not.
# Please note that this option should be used with the command
# "echo 1 > /proc/sys/vm/overcommit_memory".
# Default false
#enable_virtio_mem = true

# Disable block device from being used for a container's rootfs.
# In case of a storage driver like devicemapper where a container's
# root file system is backed by a block device, the block device is passed
# directly to the hypervisor for performance reasons.
# This flag prevents the block device from being passed to the hypervisor,
# virtio-fs is used instead to pass the rootfs.
disable_block_device_use = false

# Shared file system type:
#   - virtio-fs (default)
#   - virtio-9p
#   - virtio-fs-nydus
shared_fs = "virtio-fs"

# Path to vhost-user-fs daemon.
virtio_fs_daemon = "/usr/libexec/virtiofsd"

# List of valid annotations values for the virtiofs daemon
# The default if not set is empty (all annotations rejected.)
# Your distribution recommends: ["/usr/libexec/virtiofsd"]
valid_virtio_fs_daemon_paths = ["/usr/libexec/virtiofsd"]

# Default size of DAX cache in MiB
virtio_fs_cache_size = 0

# Extra args for virtiofsd daemon
#
# Format example:
#   ["-o", "arg1=xxx,arg2", "-o", "hello world", "--arg3=yyy"]
# Examples:
#   Set virtiofsd log level to debug : ["-o", "log_level=debug"] or ["-d"]
#
# see `virtiofsd -h` for possible options.
virtio_fs_extra_args = ["--thread-pool-size=1", "-o", "announce_submounts"]

# Cache mode:
#
#  - none
#    Metadata, data, and pathname lookup are not cached in guest. They are
#    always fetched from host and any changes are immediately pushed to host.
#
#  - auto
#    Metadata and pathname lookup cache expires after a configured amount of
#    time (default is 1 second). Data is cached while the file is open (close
#    to open consistency).
#
#  - always
#    Metadata, data, and pathname lookup are cached in guest and never expire.
virtio_fs_cache = "auto"

# Block storage driver to be used for the hypervisor in case the container
# rootfs is backed by a block device. This is virtio-scsi, virtio-blk
# or nvdimm.
block_device_driver = "virtio-scsi"

# Specifies cache-related options will be set to block devices or not.
# Default false
#block_device_cache_set = true

# Specifies cache-related options for block devices.
# Denotes whether use of O_DIRECT (bypass the host page cache) is enabled.
# Default false
#block_device_cache_direct = true

# Specifies cache-related options for block devices.
# Denotes whether flush requests for the device are ignored.
# Default false
#block_device_cache_noflush = true

# Enable iothreads (data-plane) to be used. This causes IO to be
# handled in a separate IO thread. This is currently only implemented
# for SCSI.
#
enable_iothreads = false

# Enable pre allocation of VM RAM, default false
# Enabling this will result in lower container density
# as all of the memory will be allocated and locked
# This is useful when you want to reserve all the memory
# upfront or in the cases where you want memory latencies
# to be very predictable
# Default false
#enable_mem_prealloc = true

# Enable huge pages for VM RAM, default false
# Enabling this will result in the VM memory
# being allocated using huge pages.
# This is useful when you want to use vhost-user network
# stacks within the container. This will automatically
# result in memory pre allocation
#enable_hugepages = true

# Enable vhost-user storage device, default false
# Enabling this will result in some Linux reserved block type
# major range 240-254 being chosen to represent vhost-user devices.
enable_vhost_user_store = false

# The base directory specifically used for vhost-user devices.
# Its sub-path "block" is used for block devices; "block/sockets" is
# where we expect vhost-user sockets to live; "block/devices" is where
# simulated block device nodes for vhost-user devices to live.
vhost_user_store_path = "/var/run/kata-containers/vhost-user"

# Enable vIOMMU, default false
# Enabling this will result in the VM having a vIOMMU device
# This will also add the following options to the kernel's
# command line: intel_iommu=on,iommu=pt
#enable_iommu = true

# Enable IOMMU_PLATFORM, default false
# Enabling this will result in the VM device having iommu_platform=on set
#enable_iommu_platform = true

# List of valid annotations values for the vhost user store path
# The default if not set is empty (all annotations rejected.)
# Your distribution recommends: ["/var/run/kata-containers/vhost-user"]
valid_vhost_user_store_paths = ["/var/run/kata-containers/vhost-user"]

# Enable file based guest memory support. The default is an empty string which
# will disable this feature. In the case of virtio-fs, this is enabled
# automatically and '/dev/shm' is used as the backing folder.
# This option will be ignored if VM templating is enabled.
#file_mem_backend = ""

# List of valid annotations values for the file_mem_backend annotation
# The default if not set is empty (all annotations rejected.)
# Your distribution recommends: [""]
valid_file_mem_backends = [""]

# -pflash can add image file to VM. The arguments of it should be in format
# of ["/path/to/flash0.img", "/path/to/flash1.img"]
pflashes = []

# This option changes the default hypervisor and kernel parameters
# to enable debug output where available.
#
# Default false
enable_debug = true

# Disable the customizations done in the runtime when it detects
# that it is running on top a VMM. This will result in the runtime
# behaving as it would when running on bare metal.
#
#disable_nesting_checks = true

# This is the msize used for 9p shares. It is the number of bytes
# used for 9p packet payload.
#msize_9p = 8192

# If false and nvdimm is supported, use nvdimm device to plug guest image.
# Otherwise virtio-block device is used.
#
# nvdimm is not supported when `confidential_guest = true`.
#
# Default is false
#disable_image_nvdimm = true

# VFIO devices are hotplugged on a bridge by default.
# Enable hotplugging on root bus. This may be required for devices with
# a large PCI bar, as this is a current limitation with hotplugging on
# a bridge.
# Default false
#hotplug_vfio_on_root_bus = true

# Before hot plugging a PCIe device, you need to add a pcie_root_port device.
# Use this parameter when using some large PCI bar devices, such as Nvidia GPU
# The value means the number of pcie_root_port
# This value is valid when hotplug_vfio_on_root_bus is true and machine_type is "q35"
# Default 0
#pcie_root_port = 2

# If vhost-net backend for virtio-net is not desired, set to true. Default is false, which trades off
# security (vhost-net runs ring0) for network I/O performance.
#disable_vhost_net = true

#
# Default entropy source.
# The path to a host source of entropy (including a real hardware RNG)
# /dev/urandom and /dev/random are two main options.
# Be aware that /dev/random is a blocking source of entropy.  If the host
# runs out of entropy, the VMs boot time will increase leading to get startup
# timeouts.
# The source of entropy /dev/urandom is non-blocking and provides a
# generally acceptable source of entropy. It should work well for pretty much
# all practical purposes.
#entropy_source= "/dev/urandom"

# List of valid annotations values for entropy_source
# The default if not set is empty (all annotations rejected.)
# Your distribution recommends: ["/dev/urandom","/dev/random",""]
valid_entropy_sources = ["/dev/urandom","/dev/random",""]

# Path to OCI hook binaries in the *guest rootfs*.
# This does not affect host-side hooks which must instead be added to
# the OCI spec passed to the runtime.
#
# You can create a rootfs with hooks by customizing the osbuilder scripts:
# https://github.com/kata-containers/kata-containers/tree/main/tools/osbuilder
#
# Hooks must be stored in a subdirectory of guest_hook_path according to their
# hook type, i.e. "guest_hook_path/{prestart,poststart,poststop}".
# The agent will scan these directories for executable files and add them, in
# lexicographical order, to the lifecycle of the guest container.
# Hooks are executed in the runtime namespace of the guest. See the official documentation:
# https://github.com/opencontainers/runtime-spec/blob/v1.0.1/config.md#posix-platform-hooks
# Warnings will be logged if any error is encountered while scanning for hooks,
# but it will not abort container execution.
#guest_hook_path = "/usr/share/oci/hooks"
#
# Use rx Rate Limiter to control network I/O inbound bandwidth(size in bits/sec for SB/VM).
# In Qemu, we use classful qdiscs HTB(Hierarchy Token Bucket) to discipline traffic.
# Default 0-sized value means unlimited rate.
#rx_rate_limiter_max_rate = 0
# Use tx Rate Limiter to control network I/O outbound bandwidth(size in bits/sec for SB/VM).
# In Qemu, we use classful qdiscs HTB(Hierarchy Token Bucket) and ifb(Intermediate Functional Block)
# to discipline traffic.
# Default 0-sized value means unlimited rate.
#tx_rate_limiter_max_rate = 0

# Set where to save the guest memory dump file.
# If set, when GUEST_PANICKED event occurred,
# guest memeory will be dumped to host filesystem under guest_memory_dump_path,
# This directory will be created automatically if it does not exist.
#
# The dumped file(also called vmcore) can be processed with crash or gdb.
#
# WARNING:
#   Dump guest窶冱 memory can take very long depending on the amount of guest memory
#   and use much disk space.
#guest_memory_dump_path="/var/crash/kata"

# If enable paging.
# Basically, if you want to use "gdb" rather than "crash",
# or need the guest-virtual addresses in the ELF vmcore,
# then you should enable paging.
#
# See: https://www.qemu.org/docs/master/qemu-qmp-ref.html#Dump-guest-memory for details
#guest_memory_dump_paging=false

# Enable swap in the guest. Default false.
# When enable_guest_swap is enabled, insert a raw file to the guest as the swap device
# if the swappiness of a container (set by annotation "io.katacontainers.container.resource.swappiness")
# is bigger than 0.
# The size of the swap device should be
# swap_in_bytes (set by annotation "io.katacontainers.container.resource.swap_in_bytes") - memory_limit_in_bytes.
# If swap_in_bytes is not set, the size should be memory_limit_in_bytes.
# If swap_in_bytes and memory_limit_in_bytes is not set, the size should
# be default_memory.
#enable_guest_swap = true

# use legacy serial for guest console if available and implemented for architecture. Default false
#use_legacy_serial = true

# disable applying SELinux on the VMM process (default false)
disable_selinux=false

[factory]
# VM templating support. Once enabled, new VMs are created from template
# using vm cloning. They will share the same initial kernel, initramfs and
# agent memory by mapping it readonly. It helps speeding up new container
# creation and saves a lot of memory if there are many kata containers running
# on the same host.
#
# When disabled, new VMs are created from scratch.
#
# Note: Requires "initrd=" to be set ("image=" is not supported).
#
# Default false
#enable_template = true

# Specifies the path of template.
#
# Default "/run/vc/vm/template"
#template_path = "/run/vc/vm/template"

# The number of caches of VMCache:
# unspecified or == 0   --> VMCache is disabled
# > 0                   --> will be set to the specified number
#
# VMCache is a function that creates VMs as caches before using it.
# It helps speed up new container creation.
# The function consists of a server and some clients communicating
# through Unix socket.  The protocol is gRPC in protocols/cache/cache.proto.
# The VMCache server will create some VMs and cache them by factory cache.
# It will convert the VM to gRPC format and transport it when gets
# requestion from clients.
# Factory grpccache is the VMCache client.  It will request gRPC format
# VM and convert it back to a VM.  If VMCache function is enabled,
# kata-runtime will request VM from factory grpccache when it creates
# a new sandbox.
#
# Default 0
#vm_cache_number = 0

# Specify the address of the Unix socket that is used by VMCache.
#
# Default /var/run/kata-containers/cache.sock
#vm_cache_endpoint = "/var/run/kata-containers/cache.sock"

[agent.kata]
# If enabled, make the agent display debug-level messages.
# (default: disabled)
enable_debug = true

# Enable agent tracing.
#
# If enabled, the agent will generate OpenTelemetry trace spans.
#
# Notes:
#
# - If the runtime also has tracing enabled, the agent spans will be
#   associated with the appropriate runtime parent span.
# - If enabled, the runtime will wait for the container to shutdown,
#   increasing the container shutdown time slightly.
#
# (default: disabled)
#enable_tracing = true

# Comma separated list of kernel modules and their parameters.
# These modules will be loaded in the guest kernel using modprobe(8).
# The following example can be used to load two kernel modules with parameters
#  - kernel_modules=["e1000e InterruptThrottleRate=3000,3000,3000 EEE=1", "i915 enable_ppgtt=0"]
# The first word is considered as the module name and the rest as its parameters.
# Container will not be started when:
#  * A kernel module is specified and the modprobe command is not installed in the guest
#    or it fails loading the module.
#  * The module is not available in the guest or it doesn't met the guest kernel
#    requirements, like architecture and version.
#
kernel_modules=[]

# Enable debug console.

# If enabled, user can connect guest OS running inside hypervisor
# through "kata-runtime exec <sandbox-id>" command

debug_console_enabled = true

# Agent connection dialing timeout value in seconds
# (default: 30)
#dial_timeout = 30

[runtime]
# If enabled, the runtime will log additional debug messages to the
# system log
# (default: disabled)
enable_debug = true
#
# Internetworking model
# Determines how the VM should be connected to the
# the container network interface
# Options:
#
#   - macvtap
#     Used when the Container network interface can be bridged using
#     macvtap.
#
#   - none
#     Used when customize network. Only creates a tap device. No veth pair.
#
#   - tcfilter
#     Uses tc filter rules to redirect traffic from the network interface
#     provided by plugin to a tap interface connected to the VM.
#
internetworking_model="tcfilter"

# disable guest seccomp
# Determines whether container seccomp profiles are passed to the virtual
# machine and applied by the kata agent. If set to true, seccomp is not applied
# within the guest
# (default: true)
disable_guest_seccomp=false

# disable applying SELinux on the container process
# If set to false, the type `container_t` is applied to the container process by default.
# Note: To enable guest SELinux, the guest rootfs must be CentOS that is created and built
# with `SELINUX=yes`.
# (default: true)
disable_guest_selinux=true

# Apply a custom SELinux security policy to the container process inside the VM.
# This is used when you want to apply a type other than the default `container_t`,
# so general users should not uncomment and apply it.
# (format: "user:role:type")
# Note: You cannot specify MCS policy with the label because the sensitivity levels and
# categories are determined automatically by high-level container runtimes such as containerd.
#guest_selinux_label="system_u:system_r:container_t"

# If enabled, the runtime will create opentracing.io traces and spans.
# (See https://www.jaegertracing.io/docs/getting-started).
# (default: disabled)
#enable_tracing = true

# Set the full url to the Jaeger HTTP Thrift collector.
# The default if not set will be "http://localhost:14268/api/traces"
#jaeger_endpoint = ""

# Sets the username to be used if basic auth is required for Jaeger.
#jaeger_user = ""

# Sets the password to be used if basic auth is required for Jaeger.
#jaeger_password = ""

# If enabled, the runtime will not create a network namespace for shim and hypervisor processes.
# This option may have some potential impacts to your host. It should only be used when you know what you're doing.
# `disable_new_netns` conflicts with `internetworking_model=tcfilter` and `internetworking_model=macvtap`. It works only
# with `internetworking_model=none`. The tap device will be in the host network namespace and can connect to a bridge
# (like OVS) directly.
# (default: false)
#disable_new_netns = true

# if enabled, the runtime will add all the kata processes inside one dedicated cgroup.
# The container cgroups in the host are not created, just one single cgroup per sandbox.
# The runtime caller is free to restrict or collect cgroup stats of the overall Kata sandbox.
# The sandbox cgroup path is the parent cgroup of a container with the PodSandbox annotation.
# The sandbox cgroup is constrained if there is no container type annotation.
# See: https://pkg.go.dev/github.com/kata-containers/kata-containers/src/runtime/virtcontainers#ContainerType
sandbox_cgroup_only=false

# If enabled, the runtime will attempt to determine appropriate sandbox size (memory, CPU) before booting the virtual machine. In
# this case, the runtime will not dynamically update the amount of memory and CPU in the virtual machine. This is generally helpful
# when a hardware architecture or hypervisor solutions is utilized which does not support CPU and/or memory hotplug.
# Compatibility for determining appropriate sandbox (VM) size:
# - When running with pods, sandbox sizing information will only be available if using Kubernetes >= 1.23 and containerd >= 1.6. CRI-O
#   does not yet support sandbox sizing annotations.
# - When running single containers using a tool like ctr, container sizing information will be available.
static_sandbox_resource_mgmt=false

# If specified, sandbox_bind_mounts identifieds host paths to be mounted (ro) into the sandboxes shared path.
# This is only valid if filesystem sharing is utilized. The provided path(s) will be bindmounted into the shared fs directory.
# If defaults are utilized, these mounts should be available in the guest at `/run/kata-containers/shared/containers/sandbox-mounts`
# These will not be exposed to the container workloads, and are only provided for potential guest services.
sandbox_bind_mounts=[]

# VFIO Mode
# Determines how VFIO devices should be be presented to the container.
# Options:
#
#  - vfio
#    Matches behaviour of OCI runtimes (e.g. runc) as much as
#    possible.  VFIO devices will appear in the container as VFIO
#    character devices under /dev/vfio.  The exact names may differ
#    from the host (they need to match the VM's IOMMU group numbers
#    rather than the host's)
#
#  - guest-kernel
#    This is a Kata-specific behaviour that's useful in certain cases.
#    The VFIO device is managed by whatever driver in the VM kernel
#    claims it.  This means it will appear as one or more device nodes
#    or network interfaces depending on the nature of the device.
#    Using this mode requires specially built workloads that know how
#    to locate the relevant device interfaces within the VM.
#
vfio_mode="guest-kernel"

# If enabled, the runtime will not create Kubernetes emptyDir mounts on the guest filesystem. Instead, emptyDir mounts will
# be created on the host and shared via virtio-fs. This is potentially slower, but allows sharing of files from host to guest.
disable_guest_empty_dir=false

# Enabled experimental feature list, format: ["a", "b"].
# Experimental features are features not stable enough for production,
# they may break compatibility, and are prepared for a big version bump.
# Supported experimental features:
# (default: [])
experimental=[]

# If enabled, user can run pprof tools with shim v2 process through kata-monitor.
# (default: false)
# enable_pprof = true

# WARNING: All the options in the following section have not been implemented yet.
# This section was added as a placeholder. DO NOT USE IT!
[image]
# Container image service.
#
# Offload the CRI image management service to the Kata agent.
# (default: false)
#service_offload = true

# Container image decryption keys provisioning.
# Applies only if service_offload is true.
# Keys can be provisioned locally (e.g. through a special command or
# a local file) or remotely (usually after the guest is remotely attested).
# The provision setting is a complete URL that lets the Kata agent decide
# which method to use in order to fetch the keys.
#
# Keys can be stored in a local file, in a measured and attested initrd:
#provision=data:///local/key/file
#
# Keys could be fetched through a special command or binary from the
# initrd (guest) image, e.g. a firmware call:
#provision=file:///path/to/bin/fetcher/in/guest
#
# Keys can be remotely provisioned. The Kata agent fetches them from e.g.
# a HTTPS URL:
#provision=https://my-key-broker.foo/tenant/<tenant-id>
```
</details>

[1]
kata kenel config version: 94 (latest)

[2]
kata kenel config version: 95
```
# SELinux support:
CONFIG_AUDIT=y
CONFIG_AUDITSYSCALL=y
CONFIG_LSM_MMAP_MIN_ADDR=6553
CONFIG_NETWORK_SECMARK=y
CONFIG_SECURITY_NETWORK=y
CONFIG_SECURITY_SELINUX=y
CONFIG_SECURITY_SELINUX_BOOTPARAM=y
CONFIG_SECURITY_SELINUX_DEVELOP=y
CONFIG_SECURITY_SELINUX_CHECKREQPROT_VALUE=0
CONFIG_SECURITY_SELINUX_SIDTAB_HASH_BITS=9
CONFIG_SECURITY_SELINUX_SID2STR_CACHE_SIZE=256
```

ManaSugi commented:
We can take two approaches to enable SELinux in guest.

## Type A: SELinux is enabled on the host (Current implementation)
   * Label the guest rootfs using `restorecon` and `chcon` in advance
   * Leverage the labels for the mount and the process passed from the high-level container runtime such as containerd

![type-a-selinux-enabled](https://user-images.githubusercontent.com/66765615/183407907-2b9294f8-7004-42bc-9d25-d5999eddb880.png)

## Type B: SELinux is disabled on the host
   * Label the guest rootfs using `setfattr` in advance
   * Create the label for the mount and the process by our `kata-runtime` and label the container rootfs using `setfattr`
   * In this type, the kata-runtime needs to manage SELinux label for the container instead of the high-level runtime such as containerd.

![type-b-selinux-disabled](https://user-images.githubusercontent.com/66765615/183408624-43a9ca6d-000e-4cd4-b135-2fb1aba1c828.png)

ManaSugi commented:
@fidencio Thank you for your review!

> I think the way to go is by only enabling SELinux inside the guest in case SELinux is supported n the host. That's simpler, already takes advantage of the machinery we have, and doesn't add much complexity to the runtime.

I agree with you, so we'll adopt the current implementation (type A).

> - If disable_guest_selinux = false, we should automaically:
>   - set -o xattr to virtiofsd
>     - A message should be printed to inform this happened
>   - set selinux=1 to the kernel command line
>     - A message should be printed to inform this happened
>
> About the configurations added, I would only expose those to QEMU and Cloud Hypervisor as those are the only VMMs we got working with SELinux.

It's very nice, I'll fix it to change configurations automatically if `disable_guest_selinux=false`. Then, I'll only expose them to QEMU and CH.

> I'm thinking here on how to distribute an image with SELinux and I'm sincerely considering to have, apart from the kata-deploy image we have, a second kata-deploy-selinux image, shipping then with the CentOS rootfs image and SELinux enabled by default there, but I'd like to hear from you and from others on this idea.

This proposal is very nice, but I think this would be difficult because SELinux is disabled on the CI environment and we cannot label the image using `restorecon/chcon` which requires `selinuxfs`. As a result, we cannot publish the selinux image to [the release page](https://github.com/kata-containers/kata-containers/releases) and the `kata-delploy` cannot pull it from there.
(We may label the image by `setfattr` manually like type B, but this will be a hard work.)
Could you give me your thoughts?

fidencio commented:
@ManaSugi 

> > I'm thinking here on how to distribute an image with SELinux and I'm sincerely considering to have, apart from the kata-deploy image we have, a second kata-deploy-selinux image, shipping then with the CentOS rootfs image and SELinux enabled by default there, but I'd like to hear from you and from others on this idea.
> 
> This proposal is very nice, but I think this would be difficult because SELinux is disabled on the CI environment and we cannot label the image using `restorecon/chcon` which requires `selinuxfs`. As a result, we cannot publish the selinux image to [the release page](https://github.com/kata-containers/kata-containers/releases) and the `kata-delploy` cannot pull it from there. (We may label the image by `setfattr` manually like type B on the CI environment, but this will be a hard work.) Could you give me your thoughts?

Hmm.  This touches two parts: CI and release.

For the CI part, it'd be easy to have a CentOS 8 stream job, running with SELinux, so we can actually ensure SELinux works on the host and on the guest side.  To be fair, 2 jobs, one running with CRI-O and one running with containerd.  This may need deep changes in the CI script, but I still think we should do it.

About the release, that's a problem as GitHub only provides Ubuntu runners, but we could still have an external runner using CentOS 8 Stream that's only used as part of the release.  Not so easy, but still do-able.

What do you think about these ideas?

ManaSugi commented:
@fidencio Thank you for your comment.

> For the CI part, it'd be easy to have a CentOS 8 stream job, running with SELinux, so we can actually ensure SELinux works on the host and on the guest side. To be fair, 2 jobs, one running with CRI-O and one running with containerd. This may need deep changes in the CI script, but I still think we should do it.

I'm sorry for my poor explanation. I'm not worried about our jekins jobs, wanted to confirm the Github ruuner in your second comment.  But, thank you for your clarification!

> About the release, that's a problem as GitHub only provides Ubuntu runners, but we could still have an external runner using CentOS 8 Stream that's only used as part of the release. Not so easy, but still do-able.

I have worried about the GitHub runner, but as you said, we can use an external runner and release the image for SELinux.
Thanks! I'll work on it in the next week.

ManaSugi commented:
@fidencio I'm sorry for the late update. I updated the PR and also changed the requirements on [the comment](https://github.com/kata-containers/kata-containers/pull/4813#issuecomment-1204173442).

- If `disable_guest_selinux=false`, set automaically `selinux=1` to the kernel params and `xattr` to virtiofsd.
  - If `disable_guest_selinux=true`, set automatically `selinux=0` to the kernel params.
- Fix some bugs

I'd appreciate if you could review it again.

fidencio commented:
/test

ManaSugi commented:
/retest-fc
/retest-ubuntu-clh
/retest-s390x-ubuntu

ManaSugi commented:
/retest-clh-k8s-containerd

ManaSugi commented:
/retest-clh-crio

ManaSugi commented:
/test

ManaSugi commented:
/retest-clh-k8s-containerd

ManaSugi commented:
/test

ManaSugi commented:
/test

ManaSugi commented:
/test

ManaSugi commented:
/retest-clh-k8s-containerd

fidencio commented on src/agent/rustjail/Cargo.toml:
```diff
@@ -33,6 +33,7 @@ futures = "0.3.17"
 async-trait = "0.1.31"
 inotify = "0.9.2"
 libseccomp = { version = "0.2.3", optional = true }
+xattr = "0.2.3"
```
There's a conflict happening in the line above, as libseccomp got updated.

ManaSugi commented:
/test

fidencio reviewed (approved):
lgtm, thanks @ManaSugi!

hbrueckner commented on tools/osbuilder/image-builder/image_builder.sh:
```diff
@@ -402,6 +420,27 @@ create_rootfs_image() {
+			chroot "${mount_dir}" command -v restorecon > /dev/null; then
+			mount -t selinuxfs selinuxfs $selinuxfs_path
+			chroot "${mount_dir}" restorecon -RF -e ${SELINUXFS} /
+			chroot "${mount_dir}" chcon -t container_runtime_exec_t "/usr/bin/${agent_bin}"
```
Hi @ManaSugi ,

I wonder whether we should follow-up with changing the (container) SELinux policy to have the agent binary reflected in the file contexts. The `chcon` is a temporary solution -- a local permanent solution would be with `semanage fcontext`. 

However, this is actually only required when file context information get lost or if a `restorecon` is run again to correct according to the file contexts present in the policy. wdyt?

ManaSugi replied:
@hbrueckner Thank you for the comment. That's good point.
Yes, `semanage fcontext` is better than `chcon`, but the best solution is that we'll add the `kata-agent` binary to https://github.com/containers/container-selinux/blob/main/container.fc so that we'll be no longer need to set `container_runtime_exec_t` to the `kata-agent` binary by hand in our builder script. ( All we have to do is just executing `restorecon`.)
I'll send a PR to https://github.com/containers/container-selinux in order to add it.
Untill the PR is merged and a new version of `container-selinux` package that includes the changes is distributed by CentOS, it would be better to use the `chcon` because `semanage fcontext -a -f container_runtime_exec_t ***` will fail with the follwing message when the new `container-selinux` already has the changes.
```
ValueError: File context for /var/yp(/.*)? already defined
```

> this is actually only required when file context information get lost or if a restorecon is run again to correct according to the file contexts present in the policy. wdyt?

Yes, I haven't assumed that users run `restorecon` again by hand and we can keep the labels of the guest rootfs by using `tar --selinux` and `cp --preserve=context`etc., when we release it.

@hbrueckner @fidencio 
Could you give me your thoughts?

hbrueckner replied:
Hi @ManaSugi ,

> @hbrueckner Thank you for the comment. That's good point. Yes, `semanage fcontext` is better than `chcon`, but the best solution is that we'll add the `kata-agent` binary to https://github.com/containers/container-selinux/blob/main/container.fc so that we'll be no longer need to set `container_runtime_exec_t` to the `kata-agent` binary by hand in our builder script. ( All we have to do is just executing `restorecon`.) I'll send a PR to https://github.com/containers/container-selinux in order to add it as follows:
> 
> ```
> /usr/local/s?bin/kata-agent	--	gen_context(system_u:object_r:container_runtime_exec_t,s0)
> /usr/s?bin/kata-agent	        --	gen_context(system_u:object_r:container_runtime_exec_t,s0)
> ```

This seems to be me the best solution moving forward. Please cc me on your PR for review and LGTM.

> Yes, I haven't assumed that users run `restorecon` again by hand and we can keep the labels of the guest rootfs by using `tar --selinux` and `cp --preserve=context`etc., when we release it.

I think keeping `chcon` is an interim solution until the updated `container-sepolicy` is available. I would assume that the guest rootfs is not changed directly by users. If there is some risk, let's add a remark to the limitation in the release notes/documentation.

ManaSugi replied:
@hbrueckner 

> This seems to be me the best solution moving forward. Please cc me on your PR for review and LGTM.

Awesome, I'll cc you on my PR, thank you for reviewing.

> I think keeping chcon is an interim solution until the updated container-sepolicy is available. I would assume that the guest rootfs is not changed directly by users. If there is some risk, let's add a remark to the limitation in the release notes/documentation.

That's right. I'll add a comment above the `chcon` operation in the builder script and descriptions about it to the documentation. I'll update later.

ManaSugi replied:
@hbrueckner @fidencio 
I created a PR.
https://github.com/containers/container-selinux/pull/194

ManaSugi replied:
@hbrueckner 
I'll add a comment above the chcon operation in the builder script and descriptions to the documentation.

hbrueckner commented on src/runtime/virtcontainers/kata_agent.go:
```diff
@@ -908,11 +912,34 @@ func (k *kataAgent) constrainGRPCSpec(grpcSpec *grpc.Spec, passSeccomp bool, str
-		k.Logger().Info("SELinux label from config will be applied to the hypervisor process, not the VM workload")
-		grpcSpec.Process.SelinuxLabel = ""
+		if !disableGuestSeLinux {
+			k.Logger().Info("SELinux label will be applied to the contrainer process inside guest")
```
typo: `contrainer`

ManaSugi replied:
@hbrueckner Thank you for the review.
I'll update it later with the above changes.

ManaSugi replied:
I fixed it.

ManaSugi commented:
/test

ManaSugi commented:
resolved conflicts

liubin commented:
/test

liubin commented on src/agent/rustjail/src/selinux.rs:
```diff
@@ -0,0 +1,77 @@
+            .join("attr/exec")
+    }
+
+    let mut file = OpenOptions::new().write(true).open(attr_path)?;
```
Do you think here need a `truncate(true)`?

ManaSugi replied:
@liubin Thanks for the review.
Yes, we need `truncate` option, so I added it.

ManaSugi commented:
/test
